### ISSUE TITLE
TypingError with overload(..., inline='always') and generated implementation used in a closure

### ISSUE THREAD

**Issue by ARF1** [numba/numba#5143]
With numba 0.48dev0 a TypingError occurs if an implementation is dynamically generated and used in a closure with `overload(..., inline='always').

This usage pattern occurs if multiple overloads can share a common implementation but the overloaded functions have different signatures.

```python
import os

import numba
from numba import ir, njit, types
from numba.extending import overload, register_jitable

os.environ["NUMBA_DEBUG_PRINT_AFTER"] = "ir_legalization"


def bar(x):
    """A function stub to overload"""
    pass


def generate_impl(x):
    # generate different implemenations depending on x
    # for demonstration purposes, return just a single impl

    @register_jitable(inline="always")
    def impl(x, param2=None):
        return x[0]

    return impl


@overload(bar, inline="always")
def ol_bar_tuple(x):
    impl = generate_impl(x)

    def impl_stub(x):
        return impl(x)

    return impl_stub


@njit
def foo():
    a = bar((1, 2, 3))
    return a


foo()
```

Error:
```
TypingError                               Traceback (most recent call last)
Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\inline_closurecall.py in inline_closure_call(func_ir, glbls, block, i, callee, typingctx, arg_typs, typemap, calltypes, work_list, callee_validator, replace_freevars)
    360             f_typemap, f_return_type, f_calltypes = typed_passes.type_inference_stage(
--> 361                     typingctx, callee_ir, arg_typs, None)
    362         except Exception:

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\typed_passes.py in type_inference_stage(typingctx, interp, args, return_type, locals, raise_errors)
     65 
---> 66         infer.build_constraint()
     67         infer.propagate(raise_errors=raise_errors)

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\typeinfer.py in build_constraint(self)
    937             for inst in blk.body:
--> 938                 self.constrain_statement(inst)
    939 

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\typeinfer.py in constrain_statement(self, inst)
   1273         if isinstance(inst, ir.Assign):
-> 1274             self.typeof_assign(inst)
   1275         elif isinstance(inst, ir.SetItem):

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\typeinfer.py in typeof_assign(self, inst)
   1344         elif isinstance(value, (ir.Global, ir.FreeVar)):
-> 1345             self.typeof_global(inst, inst.target, value)
   1346         elif isinstance(value, ir.Arg):

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\typeinfer.py in typeof_global(self, inst, target, gvar)
   1443         try:
-> 1444             typ = self.resolve_value_type(inst, gvar.value)
   1445         except TypingError as e:

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\typeinfer.py in resolve_value_type(self, inst, val)
   1365             msg = str(e)
-> 1366         raise TypingError(msg, loc=inst.loc)
   1367 

TypingError: Untyped global name 'impl': cannot determine Numba type of 

File "", line 28:
    def impl_stub(x):
        return impl(x)
        ^


During handling of the above exception, another exception occurred:

TypingError                               Traceback (most recent call last)
c:\Users\My User\Documents\My Project\tools\inlining_test.py in 
     37 
     38 
---> 39 foo()

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\dispatcher.py in _compile_for_args(self, *args, **kws)
    399                 e.patch_message(msg)
    400 
--> 401             error_rewrite(e, 'typing')
    402         except errors.UnsupportedError as e:
    403             # Something unsupported is present in the user code, add help info

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\dispatcher.py in error_rewrite(e, issue_type)
    342                 raise e
    343             else:
--> 344                 reraise(type(e), e, None)
    345 
    346         argtypes = []

Z:\Systemdateien\Miniconda3\envs\numba47\lib\site-packages\numba\six.py in reraise(tp, value, tb)
    666             value = tp()
    667         if value.__traceback__ is not tb:
--> 668             raise value.with_traceback(tb)
    669         raise value
    670 

TypingError: Failed in nopython mode pipeline (step: inline overloaded functions)
Untyped global name 'impl': cannot determine Numba type of 

File "", line 28:
    def impl_stub(x):
        return impl(x)
        ^
```

**Comment by esc** [comment]
@ARF1 thanks for reporting this, I can confirm that I am able to reproduce this. Not yet sure where it comes from.

**Comment by esc** [comment]
@ARF1 I am not sure what you are trying to do here and I am not sure this is possible. My guess would be, that Numba isn't able to determine the return type of `generate_impl` and this results in the error message you see.

**Comment by esc** [comment]
@ARF1 looking at this further, here is an example that might do what you want:

```
from numba import njit, types, typed
from numba.extending import overload


def bar(x):
    """A function stub to overload"""
    pass


def generate_impl(x):
    if isinstance(x, types.ListType):
        def ham(x):
            return x.pop()
        return ham
    elif isinstance(x, types.DictType):
        def spam(x):
            return x[0]
        return spam


@overload(bar, inline="always")
def ol_bar_tuple(x):
    impl = generate_impl(x)
    return impl


@njit
def foo(x):
    a = bar(x)
    return a


l = typed.List()
l.append(1)
d = typed.Dict()
d[0] = 'a'

print(foo(l))
print(foo(d))
```

As you can see, the dispatching on types is now in the `generate_impl` method. I removed the closure over the generated implementation in the overload, since that wasn't needed in this case. Does that help at all?

**Comment by ARF1** [comment]
@esc If this is not supported, it is not a problem for me. I stumbled across it while I attempted to write a minimal reproducer for #5142 and thought I would report it as well while I was at it... It was not obvious to me that this would not be supported.

The problem I am solving is if you want to overload functions that can share a common "meta-implementation":

Example in pseudo-python:
```python
def bar(x)
    """A function stub to overload"""
    pass


def barbar(x, plenty=None, of=True):
    """A function stub to overload"""
    pass


def barbarbar(x, other=False, params='test'):
    """A function stub to overload"""
    pass


def generate_impl(x, plenty=None, of=True, other=False, params='test'):
    @register_jitable(inline="always")
    def super_flexible_meta_impl(x, plenty=None, of=True, other=False, params='test'):
        if plenty is not None:
            # do some custom initialisation
        else:
            # do some standard initialisation

        # run your algorithm
        if of is True:
            # algorithm 1
        else:
            # algorithm 2

        if other is True:
            # do some custom post-processing
            return <something1>
        else:
            return <something2>

    return super_flexible_meta_impl


@overload(bar)
def bar_ovl(x):
    impl = generate_impl(x)

    def impl_stub(x):
        return impl(x)

    return impl_stub


@overload(barbar)
def barbar_ovl(x, plenty=None, of=True):
    impl = generate_impl(x, plenty=plenty, of=of)

    def impl_stub(x, plenty=None, of=True):
        return impl(x, plenty=plenty, of=of)

    return impl_stub


@overload(barbarbar)
def barbarbar_ovl(x, other=False, params='test'):
    impl = generate_impl(x, other=other, params=params)

    def impl_stub(x, other=False, params='test'):
        return impl(x, other=other, params=params)

    return impl_stub
```

Here `bar`, `barbar` and `barbarbar` can share significant code sections but all have different signatures. I could pull out all the shared code sections into new functions and write separate implementations using them but that would make it hard to read and litter the code base with boiler-plate.

The approach above actually seems to work really well with the exception of #5142. Inlining the overload itself is not something I am actually trying to do.

As I said above, I only reported this issue because I happened on it while attempting to make a minimal reproducer for #5142.

Many thanks for your help. Please feel free to close this issue if the use case in OP is not valid and this not a bug.

**Comment by stuartarchibald** [comment]
I think the issue here is that because `register_jitable` is in use in generator function, the returned implementation appears as a free variable holding an unregistered python function, for which, at present, there's no type resolution mechanism. A simple way around this would be to swap `register_jitable` for `njit`.

**Comment by stuartarchibald** [comment]
Marking as a bug, think closure caching type behaviour would be needed to fix, not sure that this is actually possible.

**Comment by stuartarchibald** [comment]
This is fixed in 0.51.2, closing.
